packages/caravan-descriptors/caravan-rs/src/config.rs

```diff
@@ -0,0 +1,1176 @@
+    /// Caravan network
+    network: CaravanNetwork,
+    /// Caravan client
+    pub client: CaravanClient,
+    /// Signing quorum
+    pub quorum: Quorum,
+    /// Extended public keys
+    pub extended_public_keys: Vec<CaravanExtendedPublicKey>,
+    /// Starting address index, always 0 when exporting
+    pub starting_address_index: u32,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Descriptor(DescriptorError),
+    InvalidNetwork { requested: Network, found: Network },
+    Generic(String),
+    Json(serde_json::Error),
+}
+
+impl std::error::Error for Error {}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::Descriptor(e) => write!(f, "Descriptor error. {}", e),
+            Error::InvalidNetwork { requested, found } => write!(
+                f,
+                "Invalid network error. requested: {}, found: {}",
+                requested, found
+            ),
+            Error::Generic(e) => write!(f, "Generic error. {}", e),
+            Error::Json(e) => write!(f, "Json error. {}", e),
+        }
+    }
+}
+
+impl From<DescriptorError> for Error {
+    fn from(e: DescriptorError) -> Self {
+        Error::Descriptor(e)
+    }
+}
+
+impl From<serde_json::Error> for Error {
+    fn from(e: serde_json::Error) -> Self {
+        Error::Json(e)
+    }
+}
+
+impl CaravanConfig {
+    /// Get the bitcoin network value
+    pub fn network(&self) -> Network {
+        match self.network {
+            CaravanNetwork::Mainnet => Network::Bitcoin,
+            CaravanNetwork::Testnet => Network::Testnet,
+        }
+    }
+    /// Get the descriptor value
+    pub fn descriptor(&self, keychain: KeychainKind) -> Result<ExtendedDescriptor, Error> {
+        let required = self.quorum.required_signers;
+        let network: Network = self.network();
+
+        let result = match self.address_type {
+            CaravanAddressType::P2sh => {
+                let keys: Vec<DescriptorKey<Legacy>> = self.descriptor_keys(keychain)?;
+                descriptor! { sh ( sortedmulti_vec(required, keys) ) }
+            }
+            CaravanAddressType::P2shP2wsh => {
+                let keys: Vec<DescriptorKey<Segwitv0>> = self.descriptor_keys(keychain)?;
+                descriptor! { sh ( wsh ( sortedmulti_vec(required, keys) ) ) }
+            }
+            CaravanAddressType::P2wsh => {
+                let keys: Vec<DescriptorKey<Segwitv0>> = self.descriptor_keys(keychain)?;
+                descriptor! { wsh ( sortedmulti_vec(required, keys) ) }
+            }
+        }
+        .map_err(Error::Descriptor);
+
+        match result {
+            Ok((d, _, n)) => {
+                if n.contains(&network) {
+                    Ok(d)
+                } else {
+                    Err(Error::InvalidNetwork {
+                        requested: network,
+                        found: *n.iter().last().expect("network"),
```

Comment:
can this panic if there are no elements in `n`? do we care?

Reply:
assuming `n` is network, I think this can be empty for the config. A descriptor doesn't require it/use it for example.